src/Instrumentation/Psr15/src/Psr15Instrumentation.php

```diff
@@ -0,0 +1,129 @@
+use Psr\Http\Server\RequestHandlerInterface;
+use Throwable;
+
+/**
+ * @psalm-suppress ArgumentTypeCoercion
+ */
+class Psr15Instrumentation
+{
+    public static function register(): void
+    {
+        $instrumentation = new CachedInstrumentation('io.opentelemetry.contrib.php.psr15');
+
+        /**
+         * Create a span for each psr-15 middleware that is executed.
+         */
+        hook(
+            MiddlewareInterface::class,
+            'process',
+            pre: static function (MiddlewareInterface $middleware, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                $span = $instrumentation->tracer()->spanBuilder(sprintf('%s::%s', $class, $function))
+                    ->setAttribute('code.function', $function)
+                    ->setAttribute('code.namespace', $class)
+                    ->setAttribute('code.filepath', $filename)
+                    ->setAttribute('code.lineno', $lineno)
+                    ->startSpan();
+
+                Context::storage()->attach($span->storeInContext(Context::getCurrent()));
+            },
+            post: static function (MiddlewareInterface $middleware, array $params, ?ResponseInterface $response, ?Throwable $exception) {
+                $scope = Context::storage()->scope();
+                if (!$scope) {
+                    return;
+                }
+                $scope->detach();
+                $span = Span::fromContext($scope->context());
+                if ($exception) {
+                    $span->recordException($exception, [TraceAttributes::EXCEPTION_ESCAPED => true]);
+                    $span->setStatus(StatusCode::STATUS_ERROR, $exception->getMessage());
+                }
+                $span->end();
+            }
+        );
+
+        /**
+         * Create a span to wrap RequestHandlerInterface::handle. The first execution is assumed to be the root span, which
+         * is stored as a request attribute which may be accessed by later hooks.
+         */
+        hook(
+            RequestHandlerInterface::class,
+            'handle',
+            pre: static function (RequestHandlerInterface $handler, array $params, string $class, string $function, ?string $filename, ?int $lineno) use ($instrumentation) {
+                $request = ($params[0] instanceof ServerRequestInterface) ? $params[0] : null;
+                if (!$request instanceof ServerRequestInterface) {
+                    //`pre` hook runs before params are checked, so this is theoretically possible. activate a non-recording
+                    //span so that the post handler has something to close
+                    $root = new NonRecordingSpan();
+                    $root->activate();
+                    Context::storage()->attach($root->storeInContext(Context::getCurrent()));
+                    return;
+                }
+                $root = $request->getAttribute(SpanInterface::class);
```

Comment:
Missing `SpanContextInterface $context` argument; should use `Span::wrap()` instead.

Do we want to replace the current span with a non recording span? I think we should still create the span (has to update `return` too): 
```suggestion
                $root = $request
                    ? $request->getAttribute(SpanInterface::class)
                    : Span::getCurrent();
```